### PR TITLE
Lightning 1.x and test loops 

### PR DESCRIPTION
**Pull request by JorisCos**
### About this PR 
This PR is a first draft on test loops using Pytorch ligthning.

### Interest 
Using lightning to perform the test step would transfer the main functionalities of the eval script to `system`. This will allow easy multiprocessing and a more consistent behavior between training and testing.

### TODO
In this first draft, some functionalities from the eval script are still missing. 

- The metrics are not saved to disk (not hard to do)

- Currently, the path of the mixtures is added to `all_metrics` I don't think it's possible to do that when using batched data without changing the outputs of the dataset (here `asteroid/data/librimix_dataset.py`)

As we are going to use Lightning 1.x we will be able to refactor the datasets and take advantage of `LightningDataModule` which will be useful for the last point.

### CODE REVIEW

File: asteroid/engine/dataloader.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+    inputs = []
+    targets = []
+    sizes = []
+    if batch is not None:
+        max_len = 0
+        for elm in batch:
+            mixture, sources = elm
+            if len(mixture) > max_len:
+                max_len = len(mixture)
```
Comment (mpariente) [inline asteroid/engine/dataloader.py:18]:
```suggestion
        max_len = max(map(lambda x: len(x[0]), batch))
```

File: asteroid/engine/dataloader.py [hunk 2]
```diff
@@ -0,0 +1,33 @@
+import torch
+from asteroid.data.librimix_dataset import LibriMix
+import os
+from torch.utils.data import DataLoader
+import torch.nn.functional as F
+
+
+def collate_fn(batch):
+
```
Comment (mpariente) [inline asteroid/engine/dataloader.py:9]: Needs a docstring explaining what it does, and a descriptive name.

File: asteroid/engine/system.py [hunk 3]
```diff
@@ -33,21 +35,27 @@ class System(pl.LightningModule):
...
+            model,
+            optimizer,
+            loss_func,
+            train_loader,
+            sample_rate,
+            val_loader=None,
+            test_loader=None,
+            metrics_list=[],
+            scheduler=None,
+            config=None,
```
Comment (mpariente) [inline asteroid/engine/system.py:48]: Please don't change the indent. It's harder to see what has been changed than if the indent isn't changed.

File: asteroid/engine/system.py [hunk 4]
```diff
@@ -33,21 +35,27 @@ class System(pl.LightningModule):
...
-        config=None,
+            self,
+            model,
+            optimizer,
+            loss_func,
+            train_loader,
+            sample_rate,
+            val_loader=None,
+            test_loader=None,
+            metrics_list=[],
```
Comment (mpariente) [inline asteroid/engine/system.py:46]: I'm not sure that we want to have this here. I don't like mixing the test loader with all the rest. 

Maybe we should have a separate method `prepare_for_test` or only use the `trainer.test(system, test_data)`?
 
I'd like @jonashaag and @mhu-coder point of view here.

File: asteroid/engine/system.py [hunk 5]
```diff
@@ -33,21 +35,27 @@ class System(pl.LightningModule):
...
-        train_loader,
-        val_loader=None,
-        scheduler=None,
-        config=None,
+            self,
+            model,
+            optimizer,
+            loss_func,
+            train_loader,
+            sample_rate,
```
Comment (mpariente) [inline asteroid/engine/system.py:43]: New positional arg in backward breaking.

File: asteroid/engine/system.py [hunk 6]
```diff
@@ -150,7 +158,41 @@ def validation_epoch_end(self, outputs):
         """
         avg_loss = torch.stack([x["val_loss"] for x in outputs]).mean()
         tensorboard_logs = {"val_loss": avg_loss}
-        return {"val_loss": avg_loss, "log": tensorboard_logs, "progress_bar": tensorboard_logs}
+        return {"val_loss": avg_loss, "log": tensorboard_logs,
+                "progress_bar": tensorboard_logs}
```
Comment (mpariente) [inline asteroid/engine/system.py:162]: Again, please don't change code that you don't edit.

File: asteroid/engine/system.py [hunk 7]
```diff
@@ -150,7 +158,41 @@ def validation_epoch_end(self, outputs):
...
+        loss, reordered_sources = self.loss_func(est_targets, targets,
+                                                 return_est=True)
+        utt_metrics = batch_metrics(
+            inputs,
+            targets,
+            reordered_sources,
+            sizes,
+            sample_rate=self.sample_rate,
+            metrics_list=self.metric_list,
+        )
```
Comment (mpariente) [inline asteroid/engine/system.py:177]: Add a comment that utt_metrics is a dict of list.

File: asteroid/metrics.py [hunk 8]
```diff
@@ -101,3 +101,40 @@ def get_metrics(
...
+
+
+def batch_metrics(inputs, targets, estimates, sizes,
+                  sample_rate=16000, metrics_list="all",
+                  average=True, compute_permutation=False,
+                  ignore_metrics_errors=False, filename=None):
+
+    mix_np = inputs.cpu().data.numpy()
+    sources_np = targets.cpu().data.numpy()
+    est_sources_np = estimates.squeeze(0).cpu().data.numpy()
```
Comment (mpariente) [inline asteroid/metrics.py:113]: Why estimates needs `squeeze(0)`? That seems weird.

File: asteroid/metrics.py [hunk 9]
```diff
@@ -101,3 +101,40 @@ def get_metrics(
...
+        results[f'input_{key}'] = []
+
+    for i in range(batch_size):
+        # Compute individual metric on unpadded sources
+        metrics = get_metrics(mix_np[i, :sizes[i][0]],
+                              sources_np[i, :, :sizes[i][0]],
+                              est_sources_np[i, :, :sizes[i][0]],
+                              sample_rate=sample_rate,
+                              metrics_list=metrics_list,
+                              average=average,
```
Comment (mpariente) [inline asteroid/metrics.py:133]: Does this work if average is `False`? I think it doesn't.

File: asteroid/metrics.py [hunk 10]
```diff
@@ -101,3 +101,40 @@ def get_metrics(
...
+                              compute_permutation=compute_permutation,
+                              ignore_metrics_errors=ignore_metrics_errors,
+                              filename=filename
```
Comment (mpariente) [inline asteroid/metrics.py:136]: I would pass those as kwargs.

File: asteroid/metrics.py [hunk 11]
```diff
@@ -101,3 +101,40 @@ def get_metrics(
         return average_arrays_in_dic(utt_metrics)
     else:
         return utt_metrics
+
+
+def batch_metrics(inputs, targets, estimates, sizes,
+                  sample_rate=16000, metrics_list="all",
+                  average=True, compute_permutation=False,
+                  ignore_metrics_errors=False, filename=None):
```
Comment (mpariente) [inline asteroid/metrics.py:109]: @jonashaag I realize that `filename` is used nowhere no..

**Review by mpariente — COMMENTED** [review]
Overall, this is a good direction, but it has few limitations that you pointed out.

Also
- There is a documentation effort to make
- You should setup `black` and `pre-commit` (see CONTRIBUTING.md), and stop editing parts of the code you don't modify :wink:

File: asteroid/engine/dataloader.py (hunk 1 above)
Comment (jonashaag) [reply asteroid/engine/dataloader.py:18]: Or `max(len(x[0] for x in batch))` which I think is a bit more readable

Comment (jonashaag) [inline asteroid/engine/dataloader.py:18]:
```suggestion
        max_len = max(len(mix for mix, _ in batch))
```

File: asteroid/engine/dataloader.py [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+    targets = []
+    sizes = []
+    if batch is not None:
+        max_len = 0
+        for elm in batch:
+            mixture, sources = elm
+            if len(mixture) > max_len:
+                max_len = len(mixture)
+        for elm in batch:
+            mixture, sources = elm
```
Comment (jonashaag) [inline asteroid/engine/dataloader.py:20]: Did you know this?

```suggestion
        for mixture, sources in batch:
```

File: asteroid/engine/system.py [hunk 13]
```diff
@@ -150,7 +158,41 @@ def validation_epoch_end(self, outputs):
...
+        return utt_metrics
+
+    def test_epoch_end(self, outputs):
```
Comment (jonashaag) [inline asteroid/engine/system.py:180]: Here it would be really good to know what kind of thing `outputs` is. From reading this method's code only I have no clue. Maybe a combination of some of the following things would be helpful:

- docstring
- using tuple unpacking rather than `[0]`, `[1]` etc, for example `foo, *bars = outputs`
- using names more descriptive than `key` and `x` ;)

File: asteroid/metrics.py (hunk 11 above)
Comment (jonashaag) [reply asteroid/metrics.py:109]: It is passed to `get_metrics` below for better error reporting

File: asteroid/engine/system.py (hunk 4 above)
Comment (jonashaag) [reply asteroid/engine/system.py:46]: Good question... how would it be done with PL? From a look at their docs it seems there is only `trainer.test()` which uses the `Metric`s that you added to your module? (Or whatever you do for logging.) With no results export (to dict, CSV, ...)?

File: asteroid/metrics.py (hunk 11 above)
Comment (mpariente) [reply asteroid/metrics.py:109]: Oh, I guess I completely misunderstood the point. I thought the filename was supposed to be a log file.. 
Now I see.

File: asteroid/engine/dataloader.py (hunk 1 above)
Comment (mpariente) [reply asteroid/engine/dataloader.py:18]: Yes, true ^^

File: asteroid/engine/system.py (hunk 13 above)
Comment (mpariente) [reply asteroid/engine/system.py:180]: I fully agree with that. It was not clear at first sight and I had to read the rest of the code to get it.

File: asteroid/engine/system.py (hunk 4 above)
Comment (mpariente) [reply asteroid/engine/system.py:46]: The test loader can be passed at test time only, all the loaders can be outside of the `LightningModule` for that matter.

**Comment by mpariente** [comment]
Should we drop this PR? 
Maybe carry the `collate_function` that pads to the longest to another PR? 
Or @JorisCos, do you want to keep working on it?

**Comment by JorisCos** [comment]
You can close it, and tell me what you wan to do with the `collate_function` but it's unused so I don't know.

**Comment by JorisCos** [comment]
You can close it, and tell me what you wan to do with the `collate_function` but it's unused so I don't know.

**Comment by mpariente** [comment]
I'll keep in mind that we have this code here, but yeah, let's close this.
